src/source/ObjcGenerator.scala

```diff
@@ -389,13 +391,28 @@ class ObjcGenerator(spec: Spec) extends Generator(spec) {
       w.wl("- (NSString *)description")
       w.braced {
         w.w(s"return ").nestedN(2) {
-          w.w("[NSString stringWithFormat:@\"<%@ %p")
+          w.w("[NSString stringWithFormat:@\"<%@ %p: dict, %@")
+          w.w(">\", self.class, self, [[self toDict] description]")
+          w.wl("];")
+        }
+      }
+      w.wl
 
-          for (f <- r.fields) w.w(s" ${idObjc.field(f.ident)}:%@")
-          w.w(">\", self.class, self")
+      w.wl("- (NSDictionary *)toDict")
+      w.braced {
+        w.wl("#define _djinni_hide_null_(_o_) ((_o_)?(_o_):([NSNull null]))")
+        w.wl("")
+        w.w(s"return ").nestedN(2) {
+          w.w("@{")
+          w.w("@\"__class_name__\": [self.class description]")
 
           for (f <- r.fields) {
-            w.w(", ")
+            w.w(", @\"")
+            w.w(idObjc.field(f.ident))
+            w.w("\": ")
+
+            w.w("_djinni_hide_null_(")
+
             f.ty.resolved.base match {
```

Comment:
I think this really does need to apply toDict recursively for nested records in order to be a complete feature.
The old code for "description" didn't need that structure in the generator, because it was done at run time.  The string formatting using %@ would call description on the subobjects.  That's not true for toDict, so I think in it's current form it's only useful as an implementation of description, or for single-level records.  I think it should be made general before being merged.